src/flow/src/compute/render/reduce.rs

```diff
@@ -0,0 +1,957 @@
+
+    err_collector.run(|| {
+        arrange.write().apply_updates(now, result_updates)?;
+        Ok(())
+    });
+
+    // Deal with output:
+
+    // 1. Read all updates that were emitted between the last time this arrangement had updates and the current time.
+    let from = arrange.read().last_compaction_time().map(|n| n + 1);
+    let from = from.unwrap_or(repr::Timestamp::MIN);
+    let output_kv = arrange.read().get_updates_in_range(from..=now);
+
+    // 2. Truncate all updates stored in arrangement within that range.
+    let run_compaction = || {
+        arrange.write().compaction_to(now)?;
+        Ok(())
+    };
+    err_collector.run(run_compaction);
+
+    // 3. Output the updates.
+    // output is concat from key and val
+    output_kv.into_iter().map(|((mut key, v), ts, diff)| {
+        key.extend(v.into_iter());
+        (key, ts, diff)
+    })
+}
+
+/// eval distinct reduce plan, output the distinct, and update the arrangement
+///
+/// invariant: it'is assumed `kv`'s time is always <= now,
+/// since it's from a Collection Bundle, where future inserts are stored in arrange
+fn reduce_distinct_subgraph(
+    arrange: &ArrangeHandler,
+    kv: impl IntoIterator<Item = KeyValDiffRow>,
+    SubgraphArg {
+        now,
+        err_collector,
+        scheduler: _,
+        send,
+    }: SubgraphArg,
+) {
+    let ret = update_reduce_distinct_arrange(arrange, kv, now, err_collector).collect_vec();
+
+    // no future updates should exist here
+    if arrange.read().get_next_update_time(&now).is_some() {
+        err_collector.push_err(
+            InternalSnafu {
+                reason: "No future updates should exist in the reduce distinct arrangement",
+            }
+            .build(),
+        );
+    }
+
+    send.give(ret);
+}
+
+/// eval accumulable reduce plan by eval aggregate function and reduce the result
+///
+/// TODO: eval distinct by adding distinct input arrangement
+///
+/// invariant: it'is assumed `kv`'s time is always <= now,
+/// since it's from a Collection Bundle, where future inserts are stored in arrange
+fn reduce_accum_subgraph(
+    arrange: &ArrangeHandler,
+    distinct_input: &Option<Vec<ArrangeHandler>>,
+    kv: impl IntoIterator<Item = KeyValDiffRow>,
+    accum_plan: &AccumulablePlan,
+    SubgraphArg {
+        now,
+        err_collector,
+        scheduler,
+        send,
+    }: SubgraphArg,
+) {
+    let AccumulablePlan {
+        full_aggrs,
+        simple_aggrs,
+        distinct_aggrs,
+    } = accum_plan;
+    let mut key2vals = BTreeMap::<Row, Vec<(Row, repr::Diff)>>::new();
+
+    for ((key, val), _tick, diff) in kv {
+        // it is assumed that value is in order of insertion
+        let vals = key2vals.entry(key).or_default();
+        vals.push((val, diff));
+    }
+
+    let mut all_updates = Vec::with_capacity(key2vals.len());
+    let mut all_outputs = Vec::with_capacity(key2vals.len());
+
+    // lock the arrange for write
+    let mut arrange = arrange.write();
```

Comment:
The lock scope includes the remaining function body, is it expected?

Reply:
> The lock scope includes the remaining function body, is it expected?

Yes, it's there to prevent potential race condition, might consider use key level locl later